Add instrumentation for HTTP.rb

ericmustin opened the pull request:
Basic Instrumentation for [http.rb](https://github.com/httprb/http) using the features middleware that underlies.

- `https://github.com/httprb/http/wiki/Logging-and-Instrumentation#instrumentation`
- `https://github.com/httprb/http/pull/499`
- `https://github.com/httprb/http/tree/4-x-stable/lib/http/features`

This PR addresses [this community card](https://github.com/DataDog/dd-trace-rb/projects/2#card-14475564) and [open issue 529](https://github.com/DataDog/dd-trace-rb/issues/529)

Essentially httprb offers middleware functionality for wrapping the request and response methods in custom methods, where we inject distributed tracing context , start/stop the spans, and add the relevant span tags metadata. 

Submitting this a bit early so I can get some feedback on whether this approach is appropriate here, I think it's much cleaner than trying to overwrite the entire `Http::Client`. This is still pretty rough as I need to make the error handling a bit more defensive and ensure the appropriate params get returned when the tracer is disabled, but getting this in early to see if this approach is appropriate.

TODO: 
- clean up implementation add more robust error handling
- tests/specs
- docs
- test against chaining/tracer disabled/edge case scenarios more thoroughly

marcotc commented on lib/ddtrace/contrib/httprb/datadog_wrap.rb:
```diff
@@ -0,0 +1,109 @@
+
+            Contrib::Analytics.set_sample_rate(datadog_span, analytics_sample_rate) if analytics_enabled?
+
+            http_method = request.try(:verb).try(:to_s).try(:upcase)
```
Could you double-check that `try` is available when running this code from an app?
Running it in our test environment might include additional dependencies that provide this method, so it wouldn't be an accurate test.

ericmustin replied:
ah, yup, will do, just force of habit

marcotc commented on lib/ddtrace/contrib/httprb/patcher.rb:
```diff
@@ -0,0 +1,38 @@
+          do_once(:httprb) do
+            begin
+              register_feature!
+              ::HTTP::Options.send(:include, Instrumentation)
```
Why not use the native [`ActiveSupport::Notifications` events](https://github.com/httprb/http/blob/f311a9b5d27042f581b4073c70fb5d2763f31a27/lib/http/features/instrumentation.rb#L32-L38)?

They have an event for the request (`start_request.http`) and one for the response (`request.http`). You'll still have to stitch them together like you are already doing, but the integration will require no monkey-patching.

ericmustin replied:
my understanding from some very brief testing of this is that even when using the active support notification events you still have to enable the `use(instrumentation: <instrumentation>)` feature as documented here https://github.com/httprb/http/wiki/Logging-and-Instrumentation#instrumentation

So monkey patching of `HTTP::Options` would still be required afaik, additionally I was concerned this approach would interfere/overwrite/enter into a race condition/etc with any other attempts by users to include the `use` `logging` or `instrumentation` features (I still have to test this) and just I figured i'd remove this possibility by going one layer lower in the stack.

I will test this out as soon as I get some cycles and document, happy to refactor this up a layer to use the canonical `instrumentation` feature if it allows multiple instrumenters and still permits http header injection for distributed tracing. Either way I do think we'll need to keep some monkey patching in there.

ericmustin replied:
I clean this up a good amount to be something close to reviewable and also tested out the behavior we discussed here re: monkey patching and active support notifications. You can find a demonstration of these details [here](https://gist.github.com/ericmustin/69dbbab70f4a74ad60b06c6602429422) (requires `activesupport` and `http` gems locally) :

https://gist.github.com/ericmustin/69dbbab70f4a74ad60b06c6602429422

As you can see by running the above `$ ruby test.rb` I get the following output

```
COMP10929:httprb_testing eric.mustin$ ruby test_app.rb 
we should see our instrumentation occur but only the last instrumentation passed
starting dd instrumentation
ending dd instrumentation
200
we won't see instrumentation occur because we haven't explicitly called it
200
```

Essentially, if we attempt to modify the existing `instrumentation` feature, due to the way the `features` method is implemented [here](https://github.com/httprb/http/blob/6e4168f04ff5130647efb26a63f8f445ba4a6de9/lib/http/options.rb#L106-L117), we simply overwrite the last instrumentation value assigned. So what would happen is that in any situation where the user is currently calling `HTTP` with something like `HTTP.use(instrumentation: <instrumentation_feature_details>)` in their application code, our instrumentation would get lost. Additionally, as mentioned previously and demonstrated above, the `ActiveSupport::Notifications` events are only generated when the user explicitly calls `.use(instrumentation: <instrumentation_feature_details>)`. There's no real global  setting to enable a feature, only to register a new one. So i do think both adding  the custom feature (so as not to interfere with any existing instrumentation) and monkey patching to inject the feature as a default option, are appropriate here. 

Let me know what you think, it's possible i've missed something.

marcotc commented on lib/ddtrace/contrib/httprb/datadog_wrap.rb:
```diff
@@ -0,0 +1,109 @@
+            return super(request) unless tracer_enabled?
+
+            tracer = datadog_configuration[:tracer]
+            datadog_span = tracer.active_span
```
Question for @delner: what's our way to ensuring that `tracer.active_span` is the trace we expect in integrations like these (with separate start and end events)?

It seems like it would find the correct one no child integration "leaks", but I'm not sure how confident we are in this.

delner replied:
I think this is a good observation @marcotc; I don't think this is a safe operation because although in theory, if everything else is working as intended it should be no problem. If a span is opened (but not closed) between `wrap_request` and `wrap_response`, this would return the wrong span.

If there's a better way to 'track' the span without leaking, we should do that instead. Using `tracer.trace` with a block is the most reliable way. I think Marco suggested the use of ActiveSupport notifications elsewhere, for which we already have a module that serves as a reliable means of doing this work for you: so we might want to consider that.

ericmustin replied:
per latest comments, refactored to monkeypatching + wrapping with `.trace`, rather than AS approach

delner commented on lib/ddtrace/contrib/httprb/patcher.rb:
```diff
@@ -0,0 +1,37 @@
+        end
+
+        def register_feature!
+          ::HTTP::Options.register_feature(:datadog_wrap, Datadog::Contrib::Httprb::DatadogWrap)
```
@ericmustin Looks like some of the latest tests for 2.0 and 2.1 don't work with this: `(/app/lib/ddtrace/contrib/httprb/patcher.rb:26:in `rescue in block in patch') Unable to apply httprb integration: undefined method `register_feature' for HTTP::Options:Class`

Is an older version of HTTP.rb running for these Ruby versions? Which don't have these? Seems likely given they're older.

ericmustin replied:
yup that's corrrect, i've moved away from this `register_feature` approach now tho (in favor of monkeypatching) , so able to support back to 2.0. i think this is good to resolve but feel free to reopen if i'm mistaken

delner commented:
As its currently implemented, does instrumentation for HTTP.rb apply out of the box when you set `use: httprb`?

ericmustin commented:
very sorry for the long delay here folks. My bandwidth has been stretched thinner than i'd like. Finally had a free day so was able to get to this.

After reviewing the feedback here from @delner : https://github.com/DataDog/dd-trace-rb/pull/853#discussion_r346636353 I decided to do a rewrite of the httprb instrumentation using a `tracer.trace` to wrap the request. It involves monkey-patching, which i think is the most robust way to ensure the span is tracked without leaking.

For context, at first I did attempt to rewrite this to use the Active Support approach suggested by @marcotc here: https://github.com/DataDog/dd-trace-rb/pull/853#discussion_r343774762 However, bc `httprb` has separate AS events for a request's `start` and `end`, it isn't a clean fit w/dd-trace-rb's AS Module, and makes *"stitching them together*" difficult, introducing many of the same issues w/span context leaking that the original PR had. Ultimately, I think in order to use AS reliably, it would involve modifying the request object to attach a private `_datadog_span` key to maintain span context, which felt hackier than simply wrapping the request via monkey-patching.

I've tried to bring this back up to a reviewable state, and up to date with master. I think I need to add some tests for analytics + distributed tracing, but given the refactor, wanted to get a review to make sure I'm on the right track. Some notes for reviewer

- It's possible I have missed some of the latest features or integration requirements (changes to sampling and measured span tags, to name a few, I wasn't sure about how to handle).  

- The tests are admittedly brittle, i also noticed this warning log `/usr/local/bundle/gems/dogstatsd-ruby-4.7.0/lib/datadog/statsd/connection.rb:12: warning: instance variable @socket not initialized
` seems to emit quite frequently with a number of tests? Not sure if that's related to this PR 

- I did have a little bit of trouble getting `span.set_error` to work as expected. Let me know if there's a more elegant way to handle setting error tags for `400/500` level responses. 

- I more or less have tried to model the instrumentation after the current `http` instrumentation. 

let me know what you think. I'm sorry again about the long delay here, definitely will make sure to prioritise any feedback to get this over the finish line.

ericmustin commented:
@marcotc gotcha, ty for taking a look. ok let me clean this up otw, and will try to review/improve testing suite as well.

ericmustin commented:
@marcotc per your feedback, made some updates here to try to bring this across the finish line to unblock [issue-529](https://github.com/DataDog/dd-trace-rb/issues/529):

-  updated the tests to check more cases on different distributed tracing and sampling priority. 
-  added documentation for enabling this instrumentation in the getting started section. 
- supporting `>=2.0.0` for this gem here as that's when they appear to have stopped support <Ruby2.0, which is also where dd-trace-rb support stops. `https://github.com/httprb/http/pull/319`
- Also updated the integration syntax to be more in line with what appears to be some safety checks on compatibility, and explicitly added a min version to the integration in `/httprb/integration.rb`.

lmk what you think, cheers!

marcotc commented on docs/GettingStarted.md:
```diff
@@ -43,6 +43,7 @@ To contribute, check out the [contribution guidelines][contribution docs] and [d
      - [Grape](#grape)
      - [GraphQL](#graphql)
      - [gRPC](#grpc)
+     - [Http.rb](#http.rb)
```
nit: http.rb seems to like their name to not be capitalized: https://github.com/httprb/http#about

ericmustin replied:
good catch, updated.

marcotc commented:
Great stuff, @ericmustin!

One last request: do you think it is possible to support multiplexing, like we do for [Net/HTTP](https://github.com/DataDog/dd-trace-rb/blob/6147b52d78f9be60b64e4e3762e586a9f86735d9/docs/GettingStarted.md#nethttp), something like:
```ruby
c.use :httprb, describes: /user-[^.]+\.example\.com/ do ...
```

delner commented:
> One last request: do you think it is possible to support multiplexing, like we do for [Net/HTTP](https://github.com/DataDog/dd-trace-rb/blob/6147b52d78f9be60b64e4e3762e586a9f86735d9/docs/GettingStarted.md#nethttp)

To this end I would recommend looking at the `Resolver` we implemented for some of the other HTTP libraries, which provides this functionality. You should be able to copy the implementation mostly from them. See #953.

ericmustin commented:
@marcotc @delner Make sense. I've taken a pass at adding multiplexing here for `http.rb` along with specs for it. Borrowed heavily from the `net/http` multiplexing implementation. Have also updated the docs, both per the feedback above and to include the split_by_domain config option. Lmk what y'all think, thanks again for taking a look!

renchap commented:
Friendly bump? I would love to have this instrumentation in the gem!

joaovitoras commented:
Is this work done? I would love to have it available

marcotc commented on docs/GettingStarted.md:
```diff
@@ -946,6 +948,35 @@ alternate_client = Demo::Echo::Service.rpc_stub_class.new(
+| `distributed_tracing` | Enables [distributed tracing](#distributed-tracing) | `true` |
+| `service_name` | Service name for `httprb` instrumentation. | `'httprb'` |
+| `split_by_domain` | Uses the request domain as the service name when set to `true`. | `false` |
+| `tracer` | `Datadog::Tracer` used to perform instrumentation. Usually you don't need to set this. | `Datadog.tracer` |
```
We've removed the `tracer` option from integrations in #1079, we can remove it here too.

ericmustin replied:
ah i see, ty. updated, had to clean up the tests and some of the `pin` code to account for this as well after bringing this up to date with master. I think it's good now tho

ericmustin commented:
@marcotc thanks for the quick look, I added the integration_spec test suite for this integration and also brought things up to date after noticing some of the deprecation warnings i was seeing after bringing this up to date with master

marcotc commented:
Anything still missing @ericmustin? The current version seems to have a complete implementation, with tests and documentation. (I noticed that this PR is still tagged as `do-not-merge/WIP`)

ericmustin commented:
@marcotc i believe it's good to go, that label was added by brett when the pr was first made, i've updated

marcotc commented:
@ericmustin a few conflicts need to be resolved, but we are good to go!
